### PR TITLE
Add Validation elements of add-group Admin Panel

### PR DESCRIPTION
**Pull request by lyzadanger**
This small PR is part of https://github.com/hypothesis/product-backlog/issues/421

This PR adds some validation and an additional field, `origins`. I'm not happy with the design of the `origins` field but after banging my head on forms for a long while I'm going to submit what I've got that actually works at the moment. Bears a larger discussion about how best to use forms and schema, but that can be postponed for the moment.

### CODE REVIEW

File: h/schemas/admin_group.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from __future__ import unicode_literals
 
 import colander
-from deform.widget import SelectWidget, TextInputWidget
+from deform.widget import SelectWidget, TextAreaWidget
```
Comment (lyzadanger) [inline h/schemas/admin_group.py]: It feels weird to me that our schemas are responsible for defining interface.

File: h/schemas/admin_group.py [hunk 2]
```diff
@@ -17,18 +17,39 @@
...
 )
 
 
+def _split_origins(value):
+    if value != colander.null:
+        return [origin.strip() for origin in value.splitlines()]
+    return value
+
+
+def user_exists_validator_factory(user_svc):
```
Comment (lyzadanger) [inline h/schemas/admin_group.py]: I designed this this way such that the schema didn't have a dependency on any services/request-related things.

File: h/schemas/admin_group.py [hunk 3]
```diff
@@ -17,18 +17,39 @@
 _ = i18n.TranslationString
 
 VALID_GROUP_TYPES = (
-    ('private', _('Private')),
     ('restricted', _('Restricted')),
     ('open', _('Open')),
 )
 
 
+def _split_origins(value):
```
Comment (lyzadanger) [inline h/schemas/admin_group.py]: This feels gross hanging out here but I'm not sure of the best way to integrate it into a schema class.

File: h/schemas/admin_group.py [hunk 4]
```diff
@@ -61,6 +82,15 @@ class CreateAdminGroupSchema(CSRFSchema):
         title=_('Description'),
         description=_('Optional group description'),
         validator=colander.Length(max=GROUP_DESCRIPTION_MAX_LENGTH),
-        widget=TextInputWidget(rows=3),
+        widget=TextAreaWidget(rows=3),
+        missing=None
+    )
+
+    origins = colander.SchemaNode(
```
Comment (lyzadanger) [inline h/schemas/admin_group.py]: I spent _forever_ trying to get this to work in a way that wasn't hacky. It should really be a `Sequence`, but I couldn't, try as I might, get `Sequence` schemas/nodes to work in our current setup as there is an implicit dependency on certain types of interface (i.e. fields). To explain this in more depth, I'd rather use voices.

Comment (lyzadanger) [reply h/schemas/admin_group.py]: Follow-up: @robertknight and I used voices and he's going to take a look.

File: h/schemas/admin_group.py [hunk 5]
```diff
@@ -17,18 +17,39 @@
...
 VALID_GROUP_TYPES = (
-    ('private', _('Private')),
     ('restricted', _('Restricted')),
     ('open', _('Open')),
 )
 
 
+def _split_origins(value):
+    if value != colander.null:
+        return [origin.strip() for origin in value.splitlines()]
```
Comment (robertknight) [inline h/schemas/admin_group.py]: This can produce empty values if the input contains empty lines. I suggest those should just be stripped out with `[... if origin.strip()]`.

File: h/schemas/admin_group.py [hunk 6]
```diff
@@ -61,6 +82,15 @@ class CreateAdminGroupSchema(CSRFSchema):
...
         description=_('Optional group description'),
         validator=colander.Length(max=GROUP_DESCRIPTION_MAX_LENGTH),
-        widget=TextInputWidget(rows=3),
+        widget=TextAreaWidget(rows=3),
+        missing=None
+    )
+
+    origins = colander.SchemaNode(
+        colander.String(),
+        title=_('Scope Origins'),
```
Comment (robertknight) [inline h/schemas/admin_group.py]: I suspect from some initial chatter in Slack that the term "Scope Origins" might be confusing. Perhaps "URL prefixes" or "Websites" might be better? We can do some "hallway testing" with eg. Heather or Arti once this is up on prod.

File: h/schemas/admin_group.py [hunk 7]
```diff
@@ -61,6 +82,15 @@ class CreateAdminGroupSchema(CSRFSchema):
...
+        widget=TextAreaWidget(rows=5),
+        hint=_('Enter scope origins (e.g. "http://www.foo.com"), one per line'),
```
Comment (robertknight) [inline h/schemas/admin_group.py]: Using the hint is great. If we could add a placeholder value to the field with an example input that would be even better.

File: tests/h/schemas/admin_group_test.py [hunk 8]
```diff
@@ -50,11 +58,69 @@ def test_it_raises_if_required_field_missing(self, group_data, bound_schema, req
...
 
         bound_schema.deserialize(group_data)
 
+    @pytest.mark.parametrize('input_origins,expected_origins', (
+        ('http://www.foo.com', ['http://www.foo.com']),
+        ('http://www.foo.com\r\nhttps://www.foo.com', ['http://www.foo.com', 'https://www.foo.com']),
+        ('http://www.foo.com   ', ['http://www.foo.com']),
+        ('http://www.foo.com\nhttps://www.foo.com', ['http://www.foo.com', 'https://www.foo.com']),
+    ))
+    def test_it_splits_origins_by_line(self, group_data, bound_schema, input_origins, expected_origins):
```
Comment (robertknight) [inline tests/h/schemas/admin_group_test.py]: Are we validating that the origin is actually an absolute URL anywhere? If not, I think we should add that. I would suggest that we might also want to make it clear somehow that only the origin part is respected at present.

File: h/schemas/admin_group.py (hunk 2 above)
Comment (robertknight) [reply h/schemas/admin_group.py]: This looks good to me. On the call yesterday I suggested moving the semantic validation up to the view level - but this has an important disadvantage that I completely forgot about: Doing validation of whether usernames, authorities exist etc. at the field level in the Colander schema results in error messages being displayed in a helpful way after submitting the form.

There is actually plenty of precedence for doing this kind of validation in the schema, but it is in `h.accounts.schemas` which I think predates the introduction of the `h.schemas` package.

So in short, I think we can keep this as-is.

**Comment by lyzadanger** [comment]
@robertknight 

> Validate the format of origin entries (absolute, HTTP URLs)

Forgot to mention when I was talking to you yesterday, but this was another point of frustration. I figure if I could get `colander.Sequence` to work, I could use `colander`'s built-in `URL` validator on the sequence's strings. 

Given the hacked string-field-that-turns-into-a-list-via-a-`preparer`-function setup for `origins`, I'm not sure how I'd "inject" the `colander.url` validator into this sequence of confusing events, nor do I really want to write my _own_ validator as there is one "waiting for me" in colander.

Material question is: should I hack together a URL validator now or should I wait to see if we (i.e. you) have some inspirations on how to make the `Sequence` thing work correctly? If we got `Sequence` right, I feel like this would fall into place, _including_ the other thing you mentioned:

> Filter out empty entries from the origins list, if that isn't already handled elsewhere

**Comment by lyzadanger** [comment]
@robertknight in Slack:

> I put up a PR to implement an initial version of a form list input (https://github.com/hypothesis/h/pull/4876). We might find it easier to get your current branch merged with its UI as-is and then rework the origin inputs later.

With that, I'm going merge this and then work on integrating Rob's changes to the sequence inputs.

**Comment by seanh** [comment]
> I think the initial UI is fine here as a starting point and that we should ship it, get actual group creation, editing & deletion working and then do some "hallway testing" (ie. get on a calls with H staff and ask them to go through the process of creating different group types as if they were setting stuff up for a publisher) in production.

I love this suggestion
